### ISSUE TITLE
Remove any usage of barrel files from core

### ISSUE THREAD

**Issue by pnicolli** [plone/volto#6507]
## PLIP (Plone Improvement Proposal)

This is a first step for #5419 and for deprecating the usage of barrel files entirely. Eventually, those files should be removed in Plone 7.

## Abstract

Barrel files create issues in code splitting and bundling, especially because we currently use an older bundler, but even modern bundlers have a hard time with those. [Several articles and examples can be found online](https://www.google.com/search?q=removing+barrel+files)

## Motivation

Solving this issue is mandatory to improve performance and eventually it is already decided that those files should not exist in Plone 7.

## Assumptions

Volto 18 should have the files for backwards compatibility, the usage of those should just be avoided.

## Proposal & Implementation

Update every import in core volto. Add a eslint rule that raises a warning for each import from barrel files. Update the docs to reflect this choice and guide the migration.

## Deliverables

for Plone 6.1 / Volto 18:
- [x] Update Volto to avoid using barrel files internally (#6509)
- [x] Add info about this deprecation in the upgrade guide (#6426)

for Plone 7:
- [ ] Fixer to help remove barrel imports from add-ons?

Initial work on a codemod for automating the migration:
https://github.com/plone/volto/blob/7b047fb4b2d71395ddba05b7e2570fcec70fc8d4/packages/volto/src/replace-barrel-imports.cjs

- [ ] Remove the unused barrel files.

## Risks

No risks involved. This has to be fully backwards compatible.

## Participants

@pnicolli 
@sneridagh 
@stevepiercy

**Comment by davisagli** [comment]
We shouldn't close this PLIP until the barrel files are actually removed in Plone 7.

**Comment by pnicolli** [comment]
@davisagli this was opened after a brief discussion at the conf and @sneridagh and I just thought about a first step to remove the usage of those files without actually removing the files. When I wrote the PLIP I just had this in mind but maybe did not clarify that properly here, so it's fine for me to close this. On the other hand, though, the actual removal of the files and/or any upgrade needed with it should also have a PLIP and I'm not sure there is one at this time.

**Comment by davisagli** [comment]
@pnicolli I updated the Deliverables section to make it clear what was already done for Plone 6.1 and what needs to be done for Plone 7.

**Comment by sneridagh** [comment]
@ichim-david I added in the description the pointer to the initial codemod script that I've been using in the Vite PR:

https://github.com/plone/volto/blob/7b047fb4b2d71395ddba05b7e2570fcec70fc8d4/packages/volto/src/replace-barrel-imports.cjs

**Comment by Shyam-Raghuwanshi** [comment]
hi @davisagli can I work on issue - Remove the unused barrel files
or should I wait for it

**Comment by stevepiercy** [comment]
@Shyam-Raghuwanshi PLIPs are not suitable for first-time contributors. Also @pnicolli is already assigned to this issue, and several other core contributors are working on it.

**Comment by sneridagh** [comment]
@davisagli @pnicolli @stevepiercy this one has been demoted from Plone 7.
We could make it happen in 19, but all the breakage that it will provide will be too much for the outcome (since all projects and add-ons would have to adapt).
I think that given the current scenario, I would say that the current state in Volto is good enough. We could enforce for 19 the ESlint rule that will raise a warning not just for core but for projects and add-ons as well.
TBH, I'd close it, and create another if we want to do the above.

**Comment by davisagli** [comment]
> this one has been demoted from Plone 7.

@sneridagh That means that Plone 7 will NOT remove barrel files, which is not what you meant to say, right?

**Comment by sneridagh** [comment]
@davisagli It won't be necessary to remove something that it won't have.

**Comment by sneridagh** [comment]
Let's use other wording then! Sorry about it :P

**Comment by pnicolli** [comment]
> We could make it happen in 19, but all the breakage that it will provide will be too much for the outcome (since all projects and add-ons would have to adapt). I think that given the current scenario, I would say that the current state in Volto is good enough. We could enforce for 19 the ESlint rule that will raise a warning not just for core but for projects and add-ons as well. TBH, I'd close it, and create another if we want to do the above.

I agree that we should close it if we don't plan to actually remove the barrel file from Volto 19. The current state of having the files for backwards compatibility but not using it ourselves in core is good enough in my opinion.
